### PR TITLE
feat(vitest): create `vi.when()`

### LINKED ISSUE #756: [Feature Request] Extend `vi.fn`
### Clear and concise description of the problem

We previously had a disscussion that it would be usefull to have a method on `vi.fn` that will defined behaviour when the function is called with this arguments.

Also does anyone else think that `vi.fn` API is too verbose? Maybe we can tighten it a bit. Would love to see suggestions and thoughts.

### Suggested solution

 For example:

Instead of this
```ts
vi.fn((arg1) => {
  if(arg1 === '3') return true
  if(arg1 === '4') return false
  return null
})
```

We can wright something like this (API is not finalized):

```ts
vi.fn()
  .mockReturn(null)
  .on('3').mockReturn(true)
  .on('4').mockReturn(false)
```

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by macarie**
### Description

Resolves #756

Todo:

- [x] Check types and code
- [ ] Solve `@todo`s
- [x] Write documentation
- [x] Add more tests

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.
- [ ] Please check [Allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to make review process faster. Note that this option is not available for repositories that are owned by Github organizations.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/integrations/mock/when.ts [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+ * @example
+ * using w = vi.when(spy)
+ *   .calledWith('hello')
+ *   .thenReturn('HELLO')
+ */
+interface When<Fn extends Procedure> extends Disposable {
+  /**
+   * Defines behavior for a specific set of arguments.
+   *
+   * Multiple behaviors can be stacked for the same arguments. They are matched last-registered first (LIFO stack, last in first out), so earlier entries act as fallbacks once later ones are exhausted.
```
Comment (macarie) [inline packages/vitest/src/integrations/mock/when.ts:159]: Should this be LIFO or FIFO? 👀

File: packages/vitest/src/integrations/mock/when.ts [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+  (value: ReturnType, options?: OnceBehaviorOptions | undefined) => CalledWithInstance<ReturnType, Fn>
+> & Record<
+  | 'thenThrow'
+  | 'thenReject',
+  (value: unknown, options?: BehaviorOptions | undefined) => CalledWithInstance<ReturnType, Fn>
+> & Record<
+  | 'thenThrowOnce'
+  | 'thenRejectOnce',
+  (value: unknown, options?: OnceBehaviorOptions | undefined) => CalledWithInstance<ReturnType, Fn>
+>
```
Comment (macarie) [inline packages/vitest/src/integrations/mock/when.ts]: todo: write these manually and add JSDocs

File: packages/expect/src/jest-expect.ts [hunk 3]
```diff
@@ -1058,6 +1058,16 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
     })
   })
 
+  def('toHaveBeenExhausted', function () {
+    const when = this._obj
+
+    this.assert(
+      when.isExhausted(),
+      'expected when chain to have been exhausted', // @todo improve message
```
Comment (macarie) [inline packages/expect/src/jest-expect.ts]: This needs a method that returns all the unused behaviors (and proper formatting), otherwise it becomes difficult to debug.

It would also be awesome to highlight un-accessible behaviors (e.g., `calledWith(...).thenReturn('a').thenReturn('b')`) for a better DX.

File: packages/vitest/src/integrations/mock/when.ts [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+          value,
+          times,
+          remaining: times,
+          called: false,
+        })
+      }
+
+      const calledWithInstance: CalledWithInstance<ScopedReturn, Fn> = ({
+        ...output,
+        thenThrow: (value, options) => {
```
Comment (macarie) [inline packages/vitest/src/integrations/mock/when.ts:396]: todo: validate options (`times >= 0`)

File: packages/expect/src/index.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ export {
 } from './jest-asymmetric-matchers'
 export { JestChaiExpect } from './jest-expect'
 export { JestExtend } from './jest-extend'
-export { addCustomEqualityTesters } from './jest-matcher-utils'
+export { addCustomEqualityTesters, getCustomEqualityTesters /* @todo */ } from './jest-matcher-utils'
```
Comment (macarie) [inline packages/expect/src/index.ts]: Is there a function that already retrieves the custom matchers and uses it?

**Review by macarie — COMMENTED** [review]
I have a small recursive package issue... I think `when` should live in `packages/spy`, but it needs `equals` from `pacakges/expect`, which depends on `packages/spy`...

Where would it make the most sense to have `when`? Because right now `expect` can't use the helper function as it lives in `pacakges/vitest` 🙃

File: packages/expect/src/index.ts (hunk 5 above)
Comment (sheremet-va) [reply packages/expect/src/index.ts]: I think we can just move it to utils/equals

File: docs/api/vi.md [hunk 6]
```diff
@@ -797,6 +797,134 @@ globalThis.IntersectionObserver === undefined
 IntersectionObserver === undefined
 ```
 
+### vi.when <Version type="experimental">5.0</Version> <Experimental /> {#vi-when}
```
Comment (sheremet-va) [inline docs/api/vi.md]: I think it doesn't need to be experimental in v5

File: packages/vitest/src/integrations/mock/when.ts [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+   *
+   * const { isExhausted, pendingBehaviors } = w.collectPendingBehaviors()
+   *
+   * expect(isExhausted).toBe(false)
+   *
+   * console.log(pendingBehaviors)
+   * // calledWith(2)
+   * //   ✗ thenReturn(42)    never called
+   */
+  '~getDiagnostics': () => {
```
Comment (sheremet-va) [inline packages/vitest/src/integrations/mock/when.ts]: I think we can use `_`. The reason why `~` is usually chosen is because it's the last in the list of public APIs, but we have `@internal` flag which removes this method from the bundle, its types are only available in this repo

File: packages/vitest/src/integrations/mock/when.ts [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+    }
+    default: {
+      (type satisfies never)
+
+      throw new Error(`vi.when: "${type}" is not a known method`)
+    }
+  }
+}
+
+function isExhausted(action: Behavior<unknown[], unknown>['actions'][number]): boolean {
```
Comment (sheremet-va) [inline packages/vitest/src/integrations/mock/when.ts]: Can we just have a separate type instead of doing archeology in `Behavior<unknown[], unknown>['actions'][number]`?

File: test/unit/test/mocking/vi-when.test.ts [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+        .thenReturn(value, { times })
+        .thenThrow(error, { times })
+        .thenResolve(value, { times })
+        .thenReject(error, { times })
+
+      const d = w['~getDiagnostics']()
+
+      expect(d.pendingBehaviors).toMatchInlineSnapshot(`
+        "calledWith("a", 0)
+          ✗ thenReturn(97, { times: 2 })                                         2 remaining (of 2)
```
Comment (sheremet-va) [inline test/unit/test/mocking/vi-when.test.ts]: Shouldn't it be `2 remaining (out of 2)`?

**Review by sheremet-va — DISMISSED** [review]
LGTM! Just some nitpicks

File: docs/api/vi.md [hunk 10]
```diff
@@ -797,6 +797,134 @@ globalThis.IntersectionObserver === undefined
...
+interface BehaviorOptions {
+  times?: number
+}
+
+function when(spy: Mock, options?: WhenOptions): When
+```
+
+Defines per-argument behaviors on a spy, replacing its implementation for the duration of the `when` chain.
+
+Call `.calledWith(...args)` on the returned object to specify which call arguments to match, then chain one or more `then*` methods to declare what the spy should return, throw, or resolve when invoked with those arguments. Arguments are matched with deep equality and support asymmetric matchers such as `expect.any()`.
```
Comment (hi-ogawa) [inline docs/api/vi.md:816]: What's the behavior of hanging `calledWith` without `thenXxx`? For example, I'm wondering if it's expected use case to only use `calledWith` to combine with `toHaveBeenExhausted` e.g.

```js
const spy = vi.fn()
const w = vi.when(spy).calledWith(1)
expect(w).not.toHaveBeenExhausted()
spy(1)
expect(w).toHaveBeenExhausted()
```

File: packages/vitest/src/integrations/mock/when.ts (hunk 1 above)
Comment (hi-ogawa) [reply packages/vitest/src/integrations/mock/when.ts:159]: The doc example doesn't read well to me, so I want FIFO.

```js
   * @example
   * // Stack behaviors: first call returns `false`, subsequent calls return `true`
   * vi.when(spy)
   *   .calledWith('darkMode')
   *   .thenReturn(true)
   *   .thenReturnOnce(false)
```

File: docs/api/vi.md (hunk 10 above)
Comment (macarie) [reply docs/api/vi.md:816]: It will use the provided fallback action/function. By default it would call the original `spy` implementation (which in this case is a noop).

There's indeed a bug with this, as the assertion would work even if it shouldn't. I wonder if it still makes sense for `toHaveBeenExhausted` to throw when there are no registered actions, as the check would be a bit more complex (checking that at least one behavior has an action) 🤔

File: packages/vitest/src/integrations/mock/when.ts (hunk 1 above)
Comment (macarie) [reply packages/vitest/src/integrations/mock/when.ts:159]: So here you would consider `thenReturnOnce` to be unreachable?

You'd write it like this?

```ts
vi.when(spy)
  .calledWith('darkMode')
  .thenReturnOnce(false)
  .thenReturn(true)
```

I wonder what others think of this, as the fluent/builder APIs I've used were all LIFO-style, but the other mock APIs do seem to be FIFO.

File: docs/api/vi.md (hunk 10 above)
Comment (sheremet-va) [reply docs/api/vi.md:816]: I think throwing if there are no registered behaviours makes sense

File: packages/vitest/src/integrations/mock/chai.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+      if (!isWhenChain(w)) {
+        throw new TypeError(
+          `${utils.inspect(w)} is not a \`vi.when\` instance`,
+        )
+      }
+
+      const diagnostics = w._getDiagnostics()
+
+      if (diagnostics.isExhausted === isNot) {
+        throw new chai.AssertionError(isNot
```
Comment (sheremet-va) [inline packages/vitest/src/integrations/mock/chai.ts]: I don't think you even need to care about `negate` if you use `this.assert` here instead of throwing an error

**Comment by macarie** [comment]
I think I fixed all the review comments. I fixed another edge-case I found with Hiroshi's incomplete chain example and added a couple of test cases, so hopefully the suite has enough coverage and will be able to catch regressions in the future.

File: packages/vitest/src/integrations/mock/when.ts (hunk 1 above)
Comment (hi-ogawa) [reply packages/vitest/src/integrations/mock/when.ts:159]: Not sure, but from my limited english interpretation, the API suggest it should naturally read like:

> when spy is called with x, then return y once, then return z (forever).

```js
vi.when(spy)
  .calledWith(x)
  .thenReturnOnce(y)
  .thenReturn(z)
```

so current order seems inverted.

Comment (hi-ogawa) [reply packages/vitest/src/integrations/mock/when.ts:159]: This is to be decided either by internal or public pole.

**Comment by sheremet-va** [comment]
It would be nice to have a new recipe now after the https://github.com/vitest-dev/vitest/pull/10226 was merged. See https://github.com/vitest-dev/vitest/issues/10381 for follow up, and https://main.vitest.dev/guide/recipes/db-transaction.html for example

Just a page explaining the use case of this feature

File: docs/guide/recipes/conditional-mocking.md
Comment (macarie) [inline docs/guide/recipes/conditional-mocking.md:1]: I have to triple check this tomorrow (examples were generated, have to check for API hallucinations).

I'm not even sure if it's a recipe or just a guide at this point, but I like the structure and that it's less abstract than what I've added previously.

File: docs/guide/recipes/conditional-mocking.md [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+  await expect(getUserById(db, 2)).resolves.toEqual({ name: 'Gracie' })
+})
+```
+
+The same approach works across all mock outcome types. Here is the full set of actions and their equivalents:
+
+| Action | Returns | Equivalent to |
+|---|---|---|
+| `thenReturn(value)` | `value` | `mockReturnValue(value)` |
+| `thenThrow(error)` | throws `error` | `mockImplementation(() => { throw error })` |
```
Comment (macarie) [inline docs/guide/recipes/conditional-mocking.md]: todo: `mockThrow` is a thing.

File: docs/guide/recipes/conditional-mocking.md [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+```
+
+This is particularly handy when you need to override a shared spy for just one branch of a test without manually calling cleanup code afterwards.
+
+## See also
+
+- [`vi.when`](/api/vi#vi-when)
+- [`toHaveBeenExhausted`](/api/expect#tohavebeenexhausted)
+- [`vi.isWhenChain`](/api/vi#vi-iswhenchain)
+- [Auto-Cleanup with `using`](/guide/recipes/explicit-resources) — `vi.when` also supports `using` for automatic spy restoration
```
Comment (macarie) [inline docs/guide/recipes/conditional-mocking.md]: todo: remove this, there's a section about it now.

File: docs/api/vi.md [hunk 14]
```diff
@@ -797,6 +797,134 @@ globalThis.IntersectionObserver === undefined
 IntersectionObserver === undefined
 ```
 
+### vi.when <Version>5.0</Version> {#vi-when}
```
Comment (sheremet-va) [inline docs/api/vi.md]:
```suggestion
### vi.when <Version>5.0.0</Version> {#vi-when}
```

File: docs/guide/recipes/conditional-mocking.md [hunk 15]
```diff
@@ -0,0 +1,274 @@
+---
+title: Conditional Mocking with vi.when | Recipes
+---
+
+# Conditional Mocking with `vi.when`
+
+<Version>5.0.0</Version>
```
Comment (sheremet-va) [inline docs/guide/recipes/conditional-mocking.md]:
```suggestion
```

I don't think we need the version here

File: docs/guide/recipes/conditional-mocking.md [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+
+  if (id === 2) {
+    return Promise.resolve({ id: 2, name: 'Gracie' })
+  }
+
+  return Promise.resolve(undefined)
+})
+```
+
+This works, but it becomes tedious because you have to write the argument-matching logic yourself. This is something that Vitest can handle for you when using the [`vi.when`](/api/vi#vi-when) API.
```
Comment (sheremet-va) [inline docs/guide/recipes/conditional-mocking.md]:
```suggestion
This works, but it becomes tedious because you have to write the argument-matching logic yourself. This is something that Vitest can handle for you when using the [`vi.when`](/api/vi#vi-when) <Version>5.0.0</Version> API.
```

File: docs/guide/recipes/conditional-mocking.md [hunk 17]
```diff
@@ -0,0 +1,274 @@
+---
+title: Conditional Mocking with vi.when | Recipes
+---
+
+# Conditional Mocking with `vi.when`
```
Comment (sheremet-va) [inline docs/guide/recipes/conditional-mocking.md:5]: Would be nice to also mention `vi.when` in "mocking functions" and just link here in the Learn section too (don't explain the API there for the sake of simplicity, just have a link at the bottom, I think - or a small mention in `::: tip`)

**Review by sheremet-va — DISMISSED** [review]
LGTM

**Review by sheremet-va — APPROVED** [review]
